semgrep-core/toy-matcher/Matcher.ml

```diff
@@ -0,0 +1,562 @@
+(*
+   A matcher operating on sequences of symbols,
+   using the same basic matching algorithm as semgrep-core.
+
+   The goals are:
+
+   - Clarify semgrep-core's matching algorithm by not having to deal with any
+     of the real-world complexity.
+   - Keep the semgrep-core implementation separate and easy to follow,
+     without functors.
+*)
+
+open Printf
+
+type pattern_atom =
+  | Any_symbol         (* match any single symbol *)
```

Comment:
Why not Any_char? Why symbol?

Reply:
Why not closer to what semgrep does and use:
type pattern =
  | Literal of char
  | Metavar of string
  | Ellipsis

instead of split in 2 with pattern_atom and pattern with those option string.
Also why do you need to capture the content of an Ellipsis?
Is it to imitate the ugly thing I do in semgrep-core to capture the range of the sequence of matched statements?

Reply:
> Why not Any_char? Why symbol?

It's because we don't really care that they are of type `char`. It could be anything.

> Why not closer to what semgrep does and use:
> type pattern =
> | Literal of char
> | Metavar of string
> | Ellipsis
>
> instead of split in 2 with pattern_atom and pattern with those option string.

I picked the list because the list type is well-known. It makes it clear that the pattern is structured as a list rather than any tree.

> Also why do you need to capture the content of an Ellipsis?
> Is it to imitate the ugly thing I do in semgrep-core to capture the range of the sequence of matched statements?

Yes. Like anything that determines the result of the function, it must be made part of the cache key.